core/src/test/scala/unit/kafka/network/SocketServerTest.scala

```diff
@@ -179,15 +179,20 @@ class SocketServerTest extends JUnitSuite {
 
   @Test
   def testSocketsCloseOnShutdown() {
+    import java.util.concurrent.TimeUnit
     // open a connection
     val plainSocket = connect(protocol = SecurityProtocol.PLAINTEXT)
+    plainSocket.setTcpNoDelay(true)
     val traceSocket = connect(protocol = SecurityProtocol.TRACE)
+    traceSocket.setTcpNoDelay(true)
     val bytes = new Array[Byte](40)
     // send a request first to make sure the connection has been picked up by the socket server
     sendRequest(plainSocket, bytes, Some(0))
     sendRequest(traceSocket, bytes, Some(0))
     processRequest(server.requestChannel)
-
+    // Sleep for 200ms to not run into TCP SLOW_START behaviour potentially masking issues in
+    // sending data below
+    TimeUnit.MILLISECONDS.sleep(200L)
```

Comment:
@ijuma 

> I can't reproduce the issue on my end when running trunk so your help is appreciated.

So no problem :) So experiments first then:

At 200ms Nagle or not doesn't matter anymore when experimenting since I cannot get to very high sample sizes.
If I set the timing sharper 10ms there I see this:

Nagle on fails eventually:
![nagle10ms](https://cloud.githubusercontent.com/assets/6490959/23576556/0d882ae4-00a9-11e7-8e1d-735fb20b6fb2.png)

Nagle off goes on forever with 10ms:

![nagleoff10ms](https://cloud.githubusercontent.com/assets/6490959/23576562/253373a6-00a9-11e7-8858-c23ce48e9765.png)

> Could you try replacing this by a while loop that reads from the connection (e.g. plainSocket.getInputStream.read) until -1 is returned?

Sure but I'd have to do it a few lines below (after the server.shutdown()). I think (know from experience) what I'll see there though will be a `SocketException` once the server closes on its end (while we are in a blocking `read()` in the test).

Still we could do this below the server shutdown and loop till we get `SocketException` (can just suppress it and go with the test then) from the server closing instead of the 10ms hard wait? :)